File: src/Sems.Api/Modules/Subscriptions/Application/SubscriptionService.cs
using Sems.Api.Modules.Subscriptions.Domain.Model;
using Sems.Api.Modules.Subscriptions.Domain.Repositories;
using Sems.Api.Modules.Subscriptions.Domain.Services;
using Sems.Api.Shared.Errors;
using Sems.Api.Shared.Events;

namespace Sems.Api.Modules.Subscriptions.Application;

/// <summary>Casos de uso del modulo de suscripciones.</summary>
public sealed class SubscriptionService
{
    private readonly IPlanRepository _plans;
    private readonly ISubscriptionRepository _subscriptions;
    private readonly SubscriptionManager _manager;
    private readonly IDomainEventBus _bus;

    public SubscriptionService(IPlanRepository plans, ISubscriptionRepository subscriptions,
        SubscriptionManager manager, IDomainEventBus bus)
    {
        _plans = plans;
        _subscriptions = subscriptions;
        _manager = manager;
        _bus = bus;
    }

    // ------------------------------------------------------------------ planes

    public Task<List<SubscriptionPlan>> ActivePlansAsync(CancellationToken ct = default) =>
        _plans.FindAllActiveAsync(ct);

    public async Task<SubscriptionPlan> PlanByIdAsync(Guid planId, CancellationToken ct = default) =>
        await _plans.FindByIdAsync(planId, ct) ?? throw AppException.NotFound("plan not found");

    // ----------------------------------------------------------- suscripciones

    public async Task<Subscription> SubscriptionByIdAsync(Guid subscriptionId,
        CancellationToken ct = default) =>
        await _subscriptions.FindByIdAsync(subscriptionId, ct)
        ?? throw AppException.NotFound("subscription not found");

    public Task<List<Subscription>> SubscriptionsByUserAsync(string userId,
        CancellationToken ct = default) => _subscriptions.FindByUserIdAsync(userId, ct);

    public async Task<Subscription> CreateAsync(string userId, Guid planId,
        string? stripeSubscriptionId, CancellationToken ct = default)
    {
        // El plan debe existir antes de cobrar nada.
        await PlanByIdAsync(planId, ct);

        var subscription = Subscription.Start(userId, planId, stripeSubscriptionId);
        await PublishChangeAsync(subscription, ct);
        return await _subscriptions.SaveAsync(subscription, ct);
    }

    public async Task<Subscription> CancelAsync(Guid subscriptionId, CancellationToken ct = default)
    {
        var subscription = await SubscriptionByIdAsync(subscriptionId, ct);
        _manager.EnsureCanCancel(subscription.Status);
        subscription.Cancel();
        await PublishChangeAsync(subscription, ct);
        return await _subscriptions.SaveAsync(subscription, ct);
    }

    public async Task<Subscription> ChangePlanAsync(Guid subscriptionId, Guid newPlanId,
        CancellationToken ct = default)
    {
        var subscription = await SubscriptionByIdAsync(subscriptionId, ct);
        _manager.EnsureCanChangePlan(subscription.Status);
        await PlanByIdAsync(newPlanId, ct);
        subscription.ChangePlan(newPlanId);
        await PublishChangeAsync(subscription, ct);
        return await _subscriptions.SaveAsync(subscription, ct);
    }

    /// <summary>Usado por el webhook de Stripe para reflejar el estado real del cobro.</summary>
    public async Task<Subscription> UpdateStatusFromStripeAsync(string stripeSubscriptionId,
        SubscriptionStatus status, CancellationToken ct = default)
    {
        var subscription = await _subscriptions.FindByStripeSubscriptionIdAsync(stripeSubscriptionId, ct)
                           ?? throw AppException.NotFound("subscription not found for stripe id");

        subscription.UpdateStatus(status);
        await PublishChangeAsync(subscription, ct);
        return await _subscriptions.SaveAsync(subscription, ct);
    }

    /// <summary>
    /// Avisa al resto del sistema del cambio de plan.
    ///
    /// <para>Antes viajaba por el topic <c>subscriptions.events</c>.</para>
    /// </summary>
    private async Task PublishChangeAsync(Subscription subscription, CancellationToken ct)
    {
        if (!Guid.TryParse(subscription.UserId, out var userGuid))
        {
            // El identificador no es un UUID: no se emite el evento, pero la
            // operacion de negocio no debe fallar por eso.
            return;
        }

        var plan = await _plans.FindByIdAsync(subscription.PlanId, ct);
        _bus.Publish(new DomainEvents.SubscriptionChanged(userGuid, subscription.SubscriptionId,
            plan?.Name ?? "unknown", subscription.Status.ToString()));
    }
}

File: src/main/java/com/sems/subscriptions/application/SubscriptionService.java
package com.sems.subscriptions.application;

import com.sems.shared.errors.AppException;
import com.sems.shared.events.DomainEventBus;
import com.sems.shared.events.DomainEvents;
import com.sems.subscriptions.domain.model.entities.Subscription;
import com.sems.subscriptions.domain.model.entities.SubscriptionPlan;
import com.sems.subscriptions.domain.model.valueobjects.SubscriptionStatus;
import com.sems.subscriptions.domain.repositories.SubscriptionRepositories.*;
import com.sems.subscriptions.domain.services.SubscriptionManager;
import java.util.List;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/** Casos de uso del modulo de suscripciones. */
@Service
@RequiredArgsConstructor
public class SubscriptionService {

    private final PlanRepository plans;
    private final SubscriptionRepository subscriptions;
    private final SubscriptionManager manager;
    private final DomainEventBus bus;

    // ---------------------------------------------------------------- planes

    @Transactional(readOnly = true)
    public List<SubscriptionPlan> activePlans() {
        return plans.findAllActive();
    }

    @Transactional(readOnly = true)
    public SubscriptionPlan planById(UUID planId) {
        return plans.findById(planId)
                .orElseThrow(() -> AppException.notFound("plan not found"));
    }

    // --------------------------------------------------------- suscripciones

    @Transactional(readOnly = true)
    public Subscription subscriptionById(UUID subscriptionId) {
        return subscriptions.findById(subscriptionId)
                .orElseThrow(() -> AppException.notFound("subscription not found"));
    }

    @Transactional(readOnly = true)
    public List<Subscription> subscriptionsByUser(String userId) {
        return subscriptions.findByUserId(userId);
    }

    @Transactional
    public Subscription create(String userId, UUID planId, String stripeSubscriptionId) {
        // El plan debe existir antes de cobrar nada.
        planById(planId);
        Subscription saved = subscriptions.save(
                Subscription.start(userId, planId, stripeSubscriptionId));
        publishChange(saved);
        return saved;
    }

    @Transactional
    public Subscription cancel(UUID subscriptionId) {
        Subscription subscription = subscriptionById(subscriptionId);
        manager.ensureCanCancel(subscription.getStatus());
        subscription.cancel();
        Subscription saved = subscriptions.save(subscription);
        publishChange(saved);
        return saved;
    }

    @Transactional
    public Subscription changePlan(UUID subscriptionId, UUID newPlanId) {
        Subscription subscription = subscriptionById(subscriptionId);
        manager.ensureCanChangePlan(subscription.getStatus());
        planById(newPlanId);
        subscription.changePlan(newPlanId);
        Subscription saved = subscriptions.save(subscription);
        publishChange(saved);
        return saved;
    }

    /** Usado por el webhook de Stripe para reflejar el estado real del cobro. */
    @Transactional
    public Subscription updateStatusFromStripe(String stripeSubscriptionId, SubscriptionStatus status) {
        Subscription subscription = subscriptions.findByStripeSubscriptionId(stripeSubscriptionId)
                .orElseThrow(() -> AppException.notFound("subscription not found for stripe id"));
        subscription.updateStatus(status);
        Subscription saved = subscriptions.save(subscription);
        publishChange(saved);
        return saved;
    }

    /**
     * Avisa al resto del sistema del cambio de plan.
     *
     * <p>Lo escucha el control de funciones por plan del frontend a traves de la
     * consulta de suscripcion. Antes viajaba por el topic
     * {@code subscriptions.events}.
     */
    private void publishChange(Subscription subscription) {
        String planName = plans.findById(subscription.getPlanId())
                .map(SubscriptionPlan::getName).orElse("unknown");
        try {
            bus.publish(new DomainEvents.SubscriptionChanged(
                    UUID.fromString(subscription.getUserId()), subscription.getSubscriptionId(),
                    planName, subscription.getStatus().name()));
        } catch (IllegalArgumentException e) {
            // El identificador de usuario no es un UUID: no se emite el evento,
            // pero la operacion de negocio no debe fallar por eso.
        }
    }
}
